circuits/core/manager.py

```diff
@@ -697,7 +697,10 @@ def _dispatcher(self, event, channels, remaining):  # noqa
                 if isinstance(value, GeneratorType):
                     event.waitingHandlers += 1
                     event.value.promise = True
-                    self.registerTask((event, value, None))
+                    task = (event, value, None)
+                    self.registerTask(task)
+                    # Process the task immediately
+                    self.processTask(*task)
```

Comment:
If we want to remain backward compatible we could add a flag `if event.process_immediately`